--- src/TerrariaNetCore/Terraria/Graphics/Light/LightMap.cs
+++ src/tModLoader/Terraria/Graphics/Light/LightMap.cs
@@ -4,9 +_,15 @@
 
 namespace Terraria.Graphics.Light;
 
-public class LightMap
+public partial class LightMap
 {
+	// TML: Colors are stored as an array of Vector4s (where A always equals 1)
+	// instead of an array of Vector3s in order to minimize overhead when
+	// writing the data to a texture.
+	// In the future, original behavior may be restored to reduce the memory
+	// footprint and get rid of any overhead in converting between vector types
+	// if FNA were to add a new Vector3 surface format.
-	private Vector3[] _colors;
+	private Vector4[] _colors;
 	private LightMaskMode[] _mask;
 	private FastRandom _random = FastRandom.CreateWithRandomSeed();
 	private const int DEFAULT_WIDTH = 203;
@@ -28,10 +_,10 @@
 
 	public Vector3 this[int x, int y] {
 		get {
-			return _colors[IndexOf(x, y)];
+			return ToVector3(_colors[IndexOf(x, y)]);
 		}
 		set {
-			_colors[IndexOf(x, y)] = value;
+			_colors[IndexOf(x, y)] = FromVector3(value);
 		}
 	}
 
@@ -43,13 +_,13 @@
 		LightDecayThroughHoney = new Vector3(0.75f, 0.7f, 0.6f) * 0.91f;
 		Width = 203;
 		Height = 203;
-		_colors = new Vector3[41209];
+		_colors = new Vector4[41209];
 		_mask = new LightMaskMode[41209];
 	}
 
 	public void GetLight(int x, int y, out Vector3 color)
 	{
-		color = _colors[IndexOf(x, y)];
+		color = ToVector3(_colors[IndexOf(x, y)]);
 	}
 
 	public LightMaskMode GetMask(int x, int y) => _mask[IndexOf(x, y)];
@@ -182,9 +_,21 @@
 
 	public void SetSize(int width, int height)
 	{
+		//TML:
+		// The width and height parameters here are max Y and max X.
+		// We need to add one to each dimension to get the correct Array Length on resize
+		// - Solxan
+		int neededSize = (width + 1) * (height + 1);
+
+		/*
 		if (width * height > _colors.Length) {
 			_colors = new Vector3[width * height];
 			_mask = new LightMaskMode[width * height];
+		}
+		*/
+		if (neededSize > _colors.Length) {
+			_colors = new Vector4[neededSize];
+			_mask = new LightMaskMode[neededSize];
 		}
 
 		Width = width;
